test.py

```diff
@@ -279,8 +279,13 @@ async def add_test(shortname, casename) -> None:
                     self.pending_test_count += 1
 
             for p in patterns:
-                if p in t:
-                    pending.add(asyncio.create_task(add_test(shortname, None)))
+                pn = p.split('::', 2)
```

Comment:
I think this change makes the code very hairy. You don't need to split the pattern if it's not set. You also don't recognize skip_pattern option. What you could be doing instead is an additional check of options.patterns *inside* PythonTestSuite::add_test, so that if the pattern is *provided (there is no point in splititng and checking if there is no pattern), then it is additionally studied for the case name. 

This would spare you from having to pass case name through entire add_test hierarchy.